### PR TITLE
Get fullcontact profile update

### PR DESCRIPTION
**Pull request by rosnovsky**
### Description

FullContact had deprecated their v2 API at the end of December 2019. This PR updates FullContact rule with the new API URL and request format. It also unifies spelling of FullContact, and added a few comments.

### References

https://dashboard.fullcontact.com/api-ref#person-api-(v2)

> We suggest you upgrade to using our Person Enrich API as this endpoint will be discontinued by the end of December 2019. 

### Testing

No changes to any of the tests, they all still pass. Validation also passes just fine.

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [X] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

### CODE REVIEW

File: src/rules/get-fullcontact-profile.js [hunk 1]
```diff
@@ -6,46 +6,50 @@
...
- * See [FullContact docs](http://www.fullcontact.com/developer/docs/) for full details.
+ * See [FullContact docs](https://dashboard.fullcontact.com/api-ref) for full details.
  *
  */
 
 function (user, context, callback) {
-  const FULLCONTACT_KEY = configuration.FULLCONTACT_KEY;
-  const SLACK_HOOK = configuration.SLACK_HOOK_URL;
+  // Get FullContact API key: https://dashboard.fullcontact.com/
+  const FULLCONTACT_KEY = 'YOUR FULLCONTACT API KEY';
```
Comment (shushen) [inline src/rules/get-fullcontact-profile.js]: Any particular reason why the support for `configuration` variables are removed here for `FULLCONTACT_KEY` and `SLACK_HOOK`?

Comment (rosnovsky) [reply src/rules/get-fullcontact-profile.js]: Honestly, I just couldn't find a way to pass these as variables from elsewhere, and it looked more explicit and direct to just have these inline "instructions" for the user of what to put where in the rule to make it work. But this is subjective, of course.

Comment (shushen) [reply src/rules/get-fullcontact-profile.js]: There is a global configuration object that is available for the rules and that's where the API keys and webhook URLs can be stored. Please see https://auth0.com/docs/rules/guides/configuration.

Comment (rosnovsky) [reply src/rules/get-fullcontact-profile.js]: Oops, missed it completely. My bad. Just pushed an update.

Comment (shushen) [reply src/rules/get-fullcontact-profile.js]: Awesome! Many thanks for the PR!

File: rules.json [hunk 2]
```diff
@@ -234,8 +234,8 @@
...
...t metadata is already there\n  if (user.user_metadata && user.user_metadata.fullcontact) return callback(null, user, context);\n\n  request.get('https://api.fullcontact.com/v3/person.enrich', {\n    'headers': {\n      'Authorization': 'Bearer ' + FULLCONTACT_KEY,\n    },\n    body: JSON.stringify({\n      \"email\": user.email\n    }),\n  }, (error, response, body) => {\n    if (error || (response && response.statusCode !== 200)) {\n\n      slack.alert({\n        channel: '#slack_channel',\n        text: 'FullContact API Error',\n        fields: {\n          error: error ? error.toString() : (response ? response.statusCode + ' ' + body : '')\n        }\n      });\n\n      // swallow FullContact api errors and just continue login\n      return callback(null, user, context);\n    }\n\n    // if we reach here, it means FullContact returned info and we'll add it to the metadata\n    user.user_metadata = user.user_metadata || {};\n    user.user_metadata.fullcontact = body;\n\n    auth0.users.updateUserMetadata(user.user_id, user.user_metadata);\n    context.idToken['https://example.com/fullcontact'] = user.user_metadata.fullcontact;\n    return callback(null, user, context);\n  });\n}"
```
Comment (shushen) [inline rules.json]: The generated JSON file needs be updated with the latest changes.

File: src/rules/get-fullcontact-profile.js [hunk 3]
```diff
@@ -6,46 +6,53 @@
...
 
-  // skip if fullcontact metadata is already there
+  // skip if FullContact metadata is already there
   if (user.user_metadata && user.user_metadata.fullcontact) return callback(null, user, context);
 
-  request.get('https://api.fullcontact.com/v2/person.json', {
-    qs: {
-      email:  user.email,
-      apiKey: FULLCONTACT_KEY
+  request.get('https://api.fullcontact.com/v3/person.enrich', {
```
Comment (shushen) [inline src/rules/get-fullcontact-profile.js]: @rosnovsky I had a second look at the FullContact V3 API and it appears to be a POST request: https://dashboard.fullcontact.com/api-ref#person-enrichment

```
curl -X POST https://api.fullcontact.com/v3/person.enrich \
-H "Authorization: Bearer {Your API Key}" \
-H "Content-Type: application/json" \
-d '{"email":"bart@fullcontact.com","twitter":"@bartlorang","fullName":"Bart Lorang"}'
```

**Review by shushen — DISMISSED** [review]

**Comment by joshcanhelp** [comment]
@rosnovsky - Is this something you're able to clean up? Or should we backlog a fix?

**Comment by rosnovsky** [comment]
@joshcanhelp Damn them github notifications, sorry, missed this one :(

I'm working on this right now.

**Comment by rosnovsky** [comment]
Fixed the request method (it's indeed POST, not GET), updated `rules.json`, updated the test file. I'm not sure how to properly check the expected header and URL in a test before returning `200` :/

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
A few changes and a question about keeping Slack in here. 

Also ... can you run `npm run format` to make sure the formatting is consistent?

Thank you for sticking with this!

File: src/rules/get-fullcontact-profile.js [hunk 4]
```diff
@@ -22,36 +28,34 @@ function getFullContactProfile(user, context, callback) {
...
-        return callback(null, user, context);
-      }
+      slack.alert({
+        channel: '#slack_channel',
+        text: 'FullContact API Error',
+        fields: {
+          error: error ? error.toString() : (response ? response.statusCode + ' ' + body : '')
+        }
+      });
+
```
Comment (joshcanhelp) [inline src/rules/get-fullcontact-profile.js:51]: A `console.log()` would help here. If an email is not found, you get a 404 and no profile.

File: src/rules/get-fullcontact-profile.js [hunk 5]
```diff
@@ -22,36 +28,34 @@ function getFullContactProfile(user, context, callback) {
...
+      // swallow FullContact api errors and just continue login
+      return callback(null, user, context);
+    }
+
+    // if we reach here, it means FullContact returned info and we'll add it to the metadata
+    user.user_metadata = user.user_metadata || {};
+    user.user_metadata.fullcontact = body;
```
Comment (joshcanhelp) [inline src/rules/get-fullcontact-profile.js:58]: Need to parse the JSON or else we just get a string. Something like:

```js
try {
  user.user_metadata.fullcontact = JSON.parse(body);
} catch (parseError) {
  console.log("Error parsing FullContact response: " + parseError.message);
  return callback(null, user, context);
}
```

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
Just tested. The call and everything works great, just one more processing step we need.

File: src/rules/get-fullcontact-profile.js [hunk 6]
```diff
@@ -6,14 +6,20 @@
...
 
-function getFullContactProfile(user, context, callback) {
+function (user, context, callback) {
+  // The following two configuration keys could be added to the Rules configuration object at https://manage.auth0.com/#/rules/
+  // See https://auth0.com/docs/rules/guides/configuration for details
+  
+  // Get FullContact API key: https://dashboard.fullcontact.com/
   const FULLCONTACT_KEY = configuration.FULLCONTACT_KEY;
+  
+  // Create a Slack app and get a hook URL: https://api.slack.com/messaging/webhooks#getting-started
```
[the remark this answers is not in the document]
Comment (rosnovsky) [reply src/rules/get-fullcontact-profile.js:22]: Makes total sense. I was wondering why is it even here, but thought it serves some purpose :)

**Comment by rosnovsky** [comment]
Phew, so:
- Reverted `rules.json` to its previous state (I think)
- Updated the rule itself with `JSON.parse` try/catch
- Removed Slack from the rule (because why was it there in the first place?)
- Updated test file: the rule now receives a string instead of an object
- Ran `format` (hopefully, it worked)

**Comment by joshcanhelp** [comment]
Thanks so much @rosnovsky! Made a few changes on my end:

- Added error handling to the API2 call
- We weren't actually saving the parsed object to the user profile
- I removed a property from the response that made the ID token very large. We still save the full response in Auth0

Thank you for all your work here! I'll get this merged in and released with the next round of Rule templates!

**Comment by joshcanhelp** [comment]
One more ... the Rule now skips if the config key we need is not present.

**Review by joshcanhelp — APPROVED** [review]

**Pull request merged**